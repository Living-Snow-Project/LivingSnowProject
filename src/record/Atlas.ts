import { AtlasDescription } from "../constants/Strings";

enum AtlasType {
  Undefined = -1,
  SnowAlgae,
  DirtOrDebris,
  Ash,
  WhiteSnow,
  MixOfAlgaeAndDirt,
  ForestOrVegetation,
  Other,
  Max,
}

type AtlasPickerItem = {
  value: AtlasType;
  label: string;
};

// specific format for RNPickerSelect
const AtlasPickerItems: AtlasPickerItem[] = [
  { value: AtlasType.SnowAlgae, label: AtlasDescription.SnowAlgae },
  { value: AtlasType.DirtOrDebris, label: AtlasDescription.DirtOrDebris },
  { value: AtlasType.Ash, label: AtlasDescription.Ash },
  { value: AtlasType.WhiteSnow, label: AtlasDescription.WhiteSnow },
  {
    value: AtlasType.MixOfAlgaeAndDirt,
    label: AtlasDescription.MixOfAlgaeAndDirt,
  },
  {
    value: AtlasType.ForestOrVegetation,
    label: AtlasDescription.ForestOrVegetation,
  },
  { value: AtlasType.Other, label: AtlasDescription.Other },
];

const getAtlasPickerItem = (type: AtlasType): AtlasPickerItem =>
  type > AtlasType.Undefined && type < AtlasType.Max
    ? AtlasPickerItems[type]
    : { value: AtlasType.Undefined, label: AtlasDescription.Undefined };

const getAllAtlasPickerItems = (): AtlasPickerItem[] => AtlasPickerItems;

export { AtlasType, getAtlasPickerItem, getAllAtlasPickerItems };
