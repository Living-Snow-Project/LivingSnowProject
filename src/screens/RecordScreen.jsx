import React from "react";
import PropTypes from "prop-types";
import RecordView from "../views/RecordView";

export default class RecordScreen extends React.Component {
  static propTypes = {
    navigation: PropTypes.object.isRequired,
  };

  render() {
    return <RecordView navigation={this.props.navigation} />;
  }
}
