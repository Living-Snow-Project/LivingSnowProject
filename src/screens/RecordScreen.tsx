import React, {
  useCallback,
  useContext,
  useEffect,
  useMemo,
  useRef,
  useState,
} from "react";
import {
  ActivityIndicator,
  Alert,
  ScrollView,
  TextInput,
  View,
} from "react-native";
import PropTypes from "prop-types";
import "react-native-get-random-values";
import { v4 as uuidv4 } from "uuid";
import KeyboardShift from "../components/KeyboardShift";
import Logger from "../lib/Logger";
import { formInputStyles } from "../styles/FormInput";
import HeaderButton from "../components/HeaderButton";
import {
  TypeSelector,
  AlgaeSizePicker,
  AlgaeColorPicker,
} from "../components/forms/TypeSelector";
import DateSelector from "../components/forms/DateSelector";
import CustomTextInput from "../components/forms/CustomTextInput";
import GpsCoordinatesInput from "../components/forms/GpsCoordinatesInput";
import AtlasSelector from "../components/forms/AtlasSelector";
import PhotoControl from "../components/PhotoControl";
import { isAtlas, isSample, recordDateFormat } from "../record/Record";
import { getAppSettings } from "../../AppSettings";
import TestIds from "../constants/TestIds";
import { Labels, Notifications, Placeholders } from "../constants/Strings";
import { RecordReducerActionsContext } from "../hooks/useRecordReducer";
import { AlgaeRecordPropType } from "../record/PropTypes";

type OffsetOperation = "add" | "subtract";

const dateWithOffset = (date: Date, op: OffsetOperation): Date => {
  if (op === "add") {
    return new Date(
      date.setMinutes(date.getMinutes() + date.getTimezoneOffset())
    );
  }

  return new Date(
    date.setMinutes(date.getMinutes() - date.getTimezoneOffset())
  );
};

type RecordScreenRouteProp = {
  params: {
    record: AlgaeRecord;
  };
};

type RecordScreenProps = {
  navigation: {
    goBack: () => void;
    setOptions: ({ headerRight }: { headerRight: () => JSX.Element }) => void;
  };
  route: RecordScreenRouteProp;
};

const defaultRouteProps: RecordScreenRouteProp = {
  params: {
    record: {
      id: -1,
      type: "Sample",
      date: dateWithOffset(new Date(), "subtract"), // YYYY-MM-DD
      latitude: 0,
      longitude: 0,
      size: "Select a size",
      color: "Select a color",
    },
  },
};

export default function RecordScreen({
  navigation,
  route = defaultRouteProps,
}: RecordScreenProps) {
  const appSettings = getAppSettings();

  // TODO: get updating\uploading from reducer
  const [updating, setUpdating] = useState<boolean>(false);
  const [uploading, setUploading] = useState<boolean>(false);
  const [editMode] = useState<boolean>(
    route?.params?.record?.id !== -1 && route?.params?.record !== undefined
  );

  const notesRef = useRef<TextInput>(null);
  const locationDescriptionRef = useRef<TextInput>(null);

  const recordReducerActionsContext = useContext(RecordReducerActionsContext);

  // data collected and sent to the service
  const [state, setState] = useState<AlgaeRecord>(
    editMode
      ? route.params.record
      : {
          ...defaultRouteProps.params.record,
          id: uuidv4(),
          name: appSettings.name ? appSettings.name : "Anonymous",
          organization: !appSettings.organization
            ? undefined
            : appSettings.organization,
        }
  );

  const [photos, setPhotos] = useState<Photo[]>(
    editMode && route.params.record.photos ? route.params.record.photos : []
  );

  const dateString = useMemo(() => recordDateFormat(state.date), [state.date]);

  // user input form validation
  const validateUserInput = () => {
    const isNumber = (value: string | number) => !Number.isNaN(Number(value));

    if (
      !state.latitude ||
      !state.longitude ||
      !isNumber(state.latitude) ||
      !isNumber(state.longitude)
    ) {
      Alert.alert(
        Notifications.invalidCoordinates.title,
        Notifications.invalidCoordinates.message
      );

      return false;
    }

    if (state.size === "Select a size") {
      Alert.alert(Notifications.invalidAlgaeSize.title);
      return false;
    }

    if (state.color === "Select a color") {
      Alert.alert(Notifications.invalidAlgaeColor.title);
      return false;
    }

    return true;
  };

  // unmodified records do not send these fields
  // so if the fields are empty during submission, do not send them
  const removeEmptyFields = useCallback((record: AlgaeRecord): AlgaeRecord => {
    const newRecord = { ...record };

    if (newRecord.type === "Sighting" || newRecord?.tubeId === "") {
      delete newRecord.tubeId;
    }

    if (newRecord?.locationDescription === "") {
      delete newRecord.locationDescription;
    }

    if (newRecord?.notes === "") {
      delete newRecord.notes;
    }

    if (newRecord?.atlasType && newRecord.atlasType === "Undefined") {
      delete newRecord.atlasType;
    }

    return newRecord;
  }, []);

  useEffect(() => {
    const RecordAction = editMode ? (
      <HeaderButton
        testID={TestIds.RecordScreen.UpdateButton}
        onPress={() => setUpdating(true)}
        iconName="save-outline"
        placement="right"
      />
    ) : (
      <HeaderButton
        testID={TestIds.RecordScreen.UploadButton}
        onPress={() => setUploading(true)}
        iconName="cloud-upload"
        placement="right"
      />
    );

    navigation.setOptions({
      headerRight: () => RecordAction,
    });
  }, []);

  // update record effect
  // TODO: get updating state from reducer
  useEffect(() => {
    if (!updating) return;

    if (!validateUserInput()) {
      setUpdating(false);
      return;
    }

    recordReducerActionsContext
      .updatePendingRecord({
        ...removeEmptyFields(state),
        photos,
      })
      .then(() => Alert.alert(Notifications.updateRecordSuccess.title))
      .catch(() =>
        Alert.alert(
          Notifications.updateRecordFailed.title,
          Notifications.updateRecordFailed.message
        )
      )
      .finally(() => {
        setUpdating(false);
        navigation.goBack();
      });
  }, [updating]);

  // upload record effect
  // TODO: get uploading state from reducer
  useEffect(() => {
    if (!uploading) return;

    if (!validateUserInput()) {
      setUploading(false);
      return;
    }

    recordReducerActionsContext
      .uploadRecord(removeEmptyFields(state), photos)
      .then(() =>
        Alert.alert(
          Notifications.uploadSuccess.title,
          Notifications.uploadSuccess.message
        )
      )
      .catch((error) => {
        Logger.Warn(
          `Failed to upload complete record: ${error.title}: ${error.message}`
        );
        Alert.alert(error.title, error.message);
      })
      .finally(() => {
        setUploading(false);
        recordReducerActionsContext.downloadRecords();
        navigation.goBack();
      });
  }, [uploading]);

  const gpsCoordinatesProps = editMode
    ? { coordinates: { latitude: state.latitude, longitude: state.longitude } }
    : {};

  return (
    <KeyboardShift>
      {() => (
        <ScrollView style={formInputStyles.container}>
          {uploading && (
            <View style={{ marginTop: 3 }}>
              <ActivityIndicator
                animating={uploading}
                size="large"
                color="#0000ff"
              />
            </View>
          )}

          {/* Sample, Sighting, Atlas, etc */}
          <TypeSelector
            type={state.type}
            setType={(type) =>
              setState((prev) => {
                let { atlasType } = prev;

                if (isAtlas(prev.type) && !isAtlas(type)) {
                  atlasType = "Undefined";
                } else if (!isAtlas(prev.type) && isAtlas(type)) {
                  atlasType = "Snow Algae";
                }

                return { ...prev, type, atlasType };
              })
            }
          />

          {/* Date of Sample, Sighting, Atlas, etc */}
          <DateSelector
            date={dateString}
            setDate={(newDate) =>
              setState((prev) => ({
                ...prev,
                date: dateWithOffset(new Date(newDate), "add"),
              }))
            }
          />

          {/* Tube Id: only show Tube Id when recording a Sample */}
          {isSample(state.type) && (
            <CustomTextInput
              defaultValue={state?.tubeId}
              description={Labels.RecordFields.TubeId}
              placeholder={Placeholders.RecordScreen.TubeId}
              maxLength={20}
              onChangeText={(tubeId) =>
                setState((prev) => ({ ...prev, tubeId }))
              }
              onSubmitEditing={() => locationDescriptionRef.current?.focus()}
            />
          )}

          <GpsCoordinatesInput
            {...gpsCoordinatesProps}
            setGpsCoordinates={(latitude, longitude) =>
              setState((prev) => ({ ...prev, latitude, longitude }))
            }
            onSubmitEditing={() => locationDescriptionRef.current?.focus()}
          />

          <AlgaeSizePicker
            size={state.size}
            setSize={(size) => {
              setState((prev) => ({ ...prev, size }));
            }}
          />

          <AlgaeColorPicker
            color={state.color}
            setColor={(color) => {
              setState((prev) => ({ ...prev, color }));
            }}
          />

          {state?.atlasType && (
            <AtlasSelector
              recordType={state.type}
              atlasType={state.atlasType}
              setAtlasType={(atlasType) =>
                setState((prev) => ({ ...prev, atlasType }))
              }
            />
          )}

          <CustomTextInput
            defaultValue={state?.locationDescription}
            description={`${Labels.RecordFields.LocationDescription} (limit 255 characters)`}
            placeholder={Placeholders.RecordScreen.LocationDescription}
            onChangeText={(locationDescription) =>
              setState((prev) => ({ ...prev, locationDescription }))
            }
            onSubmitEditing={() => notesRef.current?.focus()}
            ref={locationDescriptionRef}
          />

          <CustomTextInput
            defaultValue={state?.notes}
            description={`${Labels.RecordFields.Notes} (limit 255 characters)`}
            placeholder={Placeholders.RecordScreen.Notes}
            onChangeText={(notes) => setState((prev) => ({ ...prev, notes }))}
            ref={notesRef}
          />

          <PhotoControl
            navigation={navigation}
            photos={photos}
            onUpdatePhotos={(newPhotos) => setPhotos(newPhotos)}
          />
        </ScrollView>
      )}
    </KeyboardShift>
  );
}

RecordScreen.propTypes = {
  navigation: PropTypes.shape({
    goBack: PropTypes.func.isRequired,
    setOptions: PropTypes.func.isRequired,
  }).isRequired,
  route: PropTypes.shape({
    params: PropTypes.shape({
      record: AlgaeRecordPropType,
    }),
  }),
};

RecordScreen.defaultProps = {
  route: defaultRouteProps,
};
