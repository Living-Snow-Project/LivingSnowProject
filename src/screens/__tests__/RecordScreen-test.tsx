import React from "react";
import { Alert } from "react-native";
import { render, fireEvent } from "@testing-library/react-native";
import RecordScreen from "../RecordScreen";
import { RecordType } from "../../record/Record";
import {
  AtlasType,
  getAtlasPickerItem,
  getAllAtlasPickerItems,
} from "../../record/Atlas";
import { AppSettings, setAppSettings } from "../../../AppSettings";

const isAtlasVisible = (queryByText) => queryByText("Atlas Surface Data");
const isTubeIdVisible = (queryByText) => queryByText("Tube Id");

let renderer;
const navigation = {
  goBack: jest.fn(),
  navigate: jest.fn(),
  setOptions: jest.fn(),
};

describe("RecordScreen test suite", () => {
  beforeEach(() => {
    renderer = render(<RecordScreen navigation={navigation} />);
  });

  test("renders", () => {
    const { getByTestId } = renderer;
    fireEvent.press(getByTestId("calendar-pressable"));
    fireEvent(getByTestId("calendar"), "onDayPress", {
      dateString: "2022-01-01",
    });
    expect(renderer.toJSON()).toMatchSnapshot();
  });

  test("Sample selected", () => {
    const { queryByText } = renderer;
    expect(isAtlasVisible(queryByText)).toBeFalsy();
    expect(isTubeIdVisible(queryByText)).toBeTruthy();
  });

  test("Sighting selected", () => {
    const { getByTestId, queryByText } = renderer;
    fireEvent(
      getByTestId("record-type-picker"),
      "onValueChange",
      RecordType.Sighting
    );
    expect(isAtlasVisible(queryByText)).toBeFalsy();
    expect(isTubeIdVisible(queryByText)).toBeFalsy();
  });

  test("TubeId user input", () => {
    const { getByPlaceholderText, getByDisplayValue } = renderer;
    const tubeId = getByPlaceholderText(
      "Leave blank if the tube does not have an id"
    );
    const expected = "123-456";

    fireEvent.changeText(tubeId, expected);
    expect(getByDisplayValue(expected)).not.toBeNull();
  });

  test("GPS coordinates use Expo.Location", () => {
    const { getByTestId, queryByTestId } = renderer;
    const spy = jest.fn();
    jest
      .spyOn(Alert, "alert")
      .mockImplementationOnce((title, message, buttons) =>
        // @ts-ignore
        spy(buttons[1])
      );

    fireEvent.press(getByTestId("manual-gps-pressable"));
    expect(queryByTestId("gps-coordaintes-text-input")).toBeNull();
    expect(spy).toHaveBeenLastCalledWith({ text: "No", style: "cancel" });
  });

  test("GPS coordinates, manual with confirmation", () => {
    const { getByDisplayValue, getByTestId } = renderer;
    const testCoordinates = "123.456, -98.765";
    jest.spyOn(Alert, "alert").mockImplementationOnce(
      (title, message, buttons) =>
        // @ts-ignore
        buttons[0].onPress() // Yes button
    );

    fireEvent.press(getByTestId("manual-gps-pressable"));
    fireEvent.changeText(
      getByTestId("gps-coordinates-text-input"),
      testCoordinates
    );

    // modal shouldn't prompt again or reset coordinates
    fireEvent.press(getByTestId("gps-coordinates-text-input"));
    expect(getByDisplayValue(testCoordinates)).not.toBeNull();
  });

  test("GPS coordinates, manual without confirmation", () => {
    const testCoordinates = "123.456, -98.765";
    const spy = jest.fn();
    jest.spyOn(Alert, "alert").mockImplementationOnce(() => spy());

    setAppSettings({ showGpsWarning: false } as AppSettings);

    const { getByDisplayValue, getByTestId } = render(
      <RecordScreen navigation={navigation} />
    );

    fireEvent.press(getByTestId("manual-gps-pressable"));
    fireEvent.changeText(
      getByTestId("gps-coordinates-text-input"),
      testCoordinates
    );

    expect(getByDisplayValue(testCoordinates)).not.toBeNull();
    expect(spy).not.toBeCalled();
  });

  test("Atlas: Red Dot selected", () => {
    const { getByTestId, queryByText } = renderer;
    fireEvent(
      getByTestId("record-type-picker"),
      "onValueChange",
      RecordType.AtlasRedDot
    );
    expect(isAtlasVisible(queryByText)).toBeTruthy();
    expect(isTubeIdVisible(queryByText)).toBeFalsy();
  });

  test("Atlas: Red Dot with Sample selected", () => {
    const { getByTestId, queryByText } = renderer;
    fireEvent(
      getByTestId("record-type-picker"),
      "onValueChange",
      RecordType.AtlasRedDotWithSample
    );
    expect(isAtlasVisible(queryByText)).toBeTruthy();
    expect(isTubeIdVisible(queryByText)).toBeTruthy();
  });

  test("Atlas: Blue Dot selected", () => {
    const { getByTestId, queryByText } = renderer;
    fireEvent(
      getByTestId("record-type-picker"),
      "onValueChange",
      RecordType.AtlasBlueDot
    );
    expect(isAtlasVisible(queryByText)).toBeTruthy();
    expect(isTubeIdVisible(queryByText)).toBeFalsy();
  });

  test("Atlas: Blue Dot with Sample selected", () => {
    const { getByTestId, queryByText } = renderer;
    fireEvent(
      getByTestId("record-type-picker"),
      "onValueChange",
      RecordType.AtlasRedDotWithSample
    );
    expect(isAtlasVisible(queryByText)).toBeTruthy();
    expect(isTubeIdVisible(queryByText)).toBeTruthy();
  });

  test("Atlas: Red Dot surface selections", () => {
    const { getByDisplayValue, getByTestId, queryByText } = renderer;
    fireEvent(
      getByTestId("record-type-picker"),
      "onValueChange",
      RecordType.AtlasRedDot
    );
    expect(isAtlasVisible(queryByText)).toBeTruthy();
    expect(isTubeIdVisible(queryByText)).toBeFalsy();

    const atlasTypes = getAllAtlasPickerItems();
    atlasTypes.forEach((item) => {
      fireEvent(getByTestId("atlas-type-picker"), "onValueChange", item.value);
      expect(getByDisplayValue(item.label)).toBeTruthy();
    });
  });

  test("Atlas: Red Dot with Sample surface selections", () => {
    const { getByDisplayValue, getByTestId, queryByText } = renderer;
    fireEvent(
      getByTestId("record-type-picker"),
      "onValueChange",
      RecordType.AtlasRedDotWithSample
    );
    expect(isAtlasVisible(queryByText)).toBeTruthy();
    expect(isTubeIdVisible(queryByText)).toBeTruthy();

    const atlasTypes = [
      getAtlasPickerItem(AtlasType.SnowAlgae),
      getAtlasPickerItem(AtlasType.MixOfAlgaeAndDirt),
    ];
    atlasTypes.forEach((item) => {
      fireEvent(getByTestId("atlas-type-picker"), "onValueChange", item.value);
      expect(getByDisplayValue(item.label)).toBeTruthy();
    });
  });

  test("Atlas: Blue Dot surface selections", () => {
    const { getByDisplayValue, getByTestId, queryByText } = renderer;
    fireEvent(
      getByTestId("record-type-picker"),
      "onValueChange",
      RecordType.AtlasBlueDot
    );
    expect(isAtlasVisible(queryByText)).toBeTruthy();
    expect(isTubeIdVisible(queryByText)).toBeFalsy();

    const atlasTypes = getAllAtlasPickerItems();
    atlasTypes.forEach((item) => {
      fireEvent(getByTestId("atlas-type-picker"), "onValueChange", item.value);
      expect(getByDisplayValue(item.label)).toBeTruthy();
    });
  });

  test("Atlas: Blue Dot with Sample surface selections", () => {
    const { getByDisplayValue, getByTestId, queryByText } = renderer;
    fireEvent(
      getByTestId("record-type-picker"),
      "onValueChange",
      RecordType.AtlasBlueDotWithSample
    );
    expect(isAtlasVisible(queryByText)).toBeTruthy();
    expect(isTubeIdVisible(queryByText)).toBeTruthy();

    const atlasTypes = [
      getAtlasPickerItem(AtlasType.SnowAlgae),
      getAtlasPickerItem(AtlasType.MixOfAlgaeAndDirt),
    ];
    atlasTypes.forEach((item) => {
      fireEvent(getByTestId("atlas-type-picker"), "onValueChange", item.value);
      expect(getByDisplayValue(item.label)).toBeTruthy();
    });
  });

  test("Location Description user input", () => {
    const { getByPlaceholderText, getByDisplayValue } = renderer;
    const location = getByPlaceholderText("ie: Blue Lake, North Cascades, WA");
    const expected = "Excelsior Pass on High Divide Trail";

    fireEvent.changeText(location, expected);
    expect(getByDisplayValue(expected)).not.toBeNull();
  });

  test("Notes user input", () => {
    const { getByPlaceholderText, getByDisplayValue } = renderer;
    const notes = getByPlaceholderText("ie. algae growing on glacial ice");
    const expected = "Frozen lake in a cold place with runnels of red snow";

    fireEvent.changeText(notes, expected);
    expect(getByDisplayValue(expected)).not.toBeNull();
  });

  test.todo("Photos");
});
