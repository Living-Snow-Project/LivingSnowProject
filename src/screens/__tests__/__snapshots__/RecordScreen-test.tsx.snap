// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RecordScreen test suite renders 1`] = `
<View
  style={
    Object {
      "height": "100%",
      "left": 0,
      "position": "absolute",
      "top": 0,
      "transform": Array [
        Object {
          "translateY": 0,
        },
      ],
      "width": "100%",
    }
  }
>
  <RCTScrollView
    style={
      Object {
        "flex": 1,
        "marginTop": 1,
        "paddingHorizontal": 10,
      }
    }
  >
    <View>
      <Text
        style={
          Object {
            "fontSize": 15,
            "marginTop": 3,
          }
        }
      >
        Are you Taking a Sample or Reporting a Sighting?
      </Text>
      <View
        style={
          Array [
            Object {
              "alignSelf": "stretch",
            },
            undefined,
          ]
        }
      >
        <View
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
            }
          }
          testID="record-type-picker"
        >
          <View
            pointerEvents="box-only"
          >
            <TextInput
              allowFontScaling={true}
              editable={false}
              rejectResponderTermination={true}
              style={
                Array [
                  Object {
                    "backgroundColor": "#efefef",
                    "borderColor": "gray",
                    "borderRadius": 4,
                    "borderWidth": 1,
                    "minHeight": "8%",
                    "paddingHorizontal": 15,
                    "paddingVertical": 15,
                  },
                  Object {},
                ]
              }
              testID="text_input"
              underlineColorAndroid="transparent"
              value="I'm Taking a Sample"
            />
          </View>
        </View>
        <Modal
          hardwareAccelerated={false}
          onOrientationChange={[Function]}
          supportedOrientations={
            Array [
              "portrait",
              "landscape",
            ]
          }
          testID="ios_modal"
          transparent={true}
          visible={false}
        >
          <View
            accessible={true}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "flex": 1,
                "opacity": 1,
              }
            }
            testID="ios_modal_top"
          />
          <View
            style={
              Array [
                Object {
                  "alignItems": "center",
                  "backgroundColor": "#f8f8f8",
                  "borderTopColor": "#dedede",
                  "borderTopWidth": 1,
                  "flexDirection": "row",
                  "height": 45,
                  "justifyContent": "space-between",
                  "paddingHorizontal": 10,
                  "zIndex": 2,
                },
                undefined,
              ]
            }
            testID="input_accessory_view"
          >
            <View
              style={
                Array [
                  Object {
                    "flexDirection": "row",
                  },
                  undefined,
                ]
              }
            >
              <View
                accessible={true}
                focusable={true}
                onClick={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  Object {
                    "opacity": 1,
                  }
                }
              >
                <View
                  style={
                    Array [
                      Object {
                        "backgroundColor": "transparent",
                        "borderColor": "#a1a1a1",
                        "borderRightWidth": 1.5,
                        "borderTopWidth": 1.5,
                        "height": 15,
                        "width": 15,
                      },
                      undefined,
                      Object {
                        "marginLeft": 11,
                        "transform": Array [
                          Object {
                            "translateY": 4,
                          },
                          Object {
                            "rotate": "-45deg",
                          },
                        ],
                      },
                      undefined,
                      Object {},
                    ]
                  }
                />
              </View>
              <View
                accessible={true}
                focusable={true}
                onClick={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  Object {
                    "opacity": 1,
                  }
                }
              >
                <View
                  style={
                    Array [
                      Object {
                        "backgroundColor": "transparent",
                        "borderColor": "#a1a1a1",
                        "borderRightWidth": 1.5,
                        "borderTopWidth": 1.5,
                        "height": 15,
                        "width": 15,
                      },
                      undefined,
                      Object {
                        "marginLeft": 22,
                        "transform": Array [
                          Object {
                            "translateY": -5,
                          },
                          Object {
                            "rotate": "135deg",
                          },
                        ],
                      },
                      undefined,
                      Object {},
                    ]
                  }
                />
              </View>
            </View>
            <View
              accessible={true}
              focusable={true}
              hitSlop={
                Object {
                  "bottom": 4,
                  "left": 4,
                  "right": 4,
                  "top": 4,
                }
              }
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Object {
                  "opacity": 1,
                }
              }
              testID="done_button"
            >
              <View
                testID="needed_for_touchable"
              >
                <Text
                  allowFontScaling={false}
                  style={
                    Array [
                      Object {
                        "color": "#007aff",
                        "fontSize": 17,
                        "fontWeight": "600",
                        "paddingRight": 11,
                        "paddingTop": 1,
                      },
                      undefined,
                      Object {},
                    ]
                  }
                  testID="done_text"
                >
                  Done
                </Text>
              </View>
            </View>
          </View>
          <View
            style={
              Array [
                Object {
                  "backgroundColor": "#d0d4da",
                  "justifyContent": "center",
                },
                Object {
                  "height": 215,
                },
                undefined,
              ]
            }
          >
            <View>
              <RNCPicker
                items={
                  Array [
                    Object {
                      "label": "I'm Taking a Sample",
                      "testID": undefined,
                      "textColor": undefined,
                      "value": 0,
                    },
                    Object {
                      "label": "I'm Reporting a Sighting",
                      "testID": undefined,
                      "textColor": undefined,
                      "value": 1,
                    },
                    Object {
                      "label": "Atlas: Red Dot",
                      "testID": undefined,
                      "textColor": undefined,
                      "value": 2,
                    },
                    Object {
                      "label": "Atlas: Red Dot with Sample",
                      "testID": undefined,
                      "textColor": undefined,
                      "value": 3,
                    },
                    Object {
                      "label": "Atlas: Blue Dot",
                      "testID": undefined,
                      "textColor": undefined,
                      "value": 4,
                    },
                    Object {
                      "label": "Atlas: Blue Dot with Sample",
                      "testID": undefined,
                      "textColor": undefined,
                      "value": 5,
                    },
                  ]
                }
                onChange={[Function]}
                selectedIndex={0}
                style={
                  Array [
                    Object {
                      "height": 216,
                    },
                    undefined,
                  ]
                }
                testID="ios_picker"
              />
            </View>
          </View>
        </Modal>
      </View>
      <Text
        style={
          Object {
            "fontSize": 15,
            "marginTop": 3,
          }
        }
      >
        Date
      </Text>
      <View
        accessible={true}
        collapsable={false}
        focusable={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        testID="calendar-pressable"
      >
        <Text
          style={
            Object {
              "backgroundColor": "#efefef",
              "borderColor": "gray",
              "borderRadius": 4,
              "borderWidth": 1,
              "minHeight": "8%",
              "paddingHorizontal": 15,
              "paddingVertical": 15,
            }
          }
        >
          2022-01-01
        </Text>
      </View>
      <Text
        style={
          Object {
            "fontSize": 15,
            "marginTop": 3,
          }
        }
      >
        Tube Id
      </Text>
      <TextInput
        allowFontScaling={true}
        blurOnSubmit={true}
        maxLength={20}
        multiline={true}
        onChangeText={[Function]}
        onContentSizeChange={[Function]}
        onSubmitEditing={[Function]}
        placeholder="Leave blank if the tube does not have an id"
        rejectResponderTermination={true}
        returnKeyType="done"
        style={
          Array [
            Object {
              "backgroundColor": "#efefef",
              "borderColor": "gray",
              "borderRadius": 4,
              "borderWidth": 1,
              "minHeight": "8%",
              "paddingHorizontal": 15,
              "paddingVertical": 15,
            },
            Object {
              "paddingTop": 15,
            },
          ]
        }
        underlineColorAndroid="transparent"
        value=""
      />
      <Text
        style={
          Object {
            "fontSize": 15,
            "marginTop": 3,
          }
        }
      >
        GPS Coordinates (latitude, longitude)
      </Text>
      <View
        accessible={true}
        collapsable={false}
        focusable={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        testID="manual-gps-pressable"
      >
        <View
          pointerEvents="none"
        >
          <TextInput
            allowFontScaling={true}
            defaultValue=""
            editable={false}
            maxLength={30}
            rejectResponderTermination={true}
            returnKeyType="done"
            style={
              Object {
                "backgroundColor": "#efefef",
                "borderColor": "gray",
                "borderRadius": 4,
                "borderWidth": 1,
                "minHeight": "8%",
                "paddingHorizontal": 15,
                "paddingVertical": 15,
              }
            }
            underlineColorAndroid="transparent"
          />
        </View>
      </View>
      <Text
        style={
          Object {
            "fontSize": 15,
            "marginTop": 3,
          }
        }
      >
        Location Description (limit 255 characters)
      </Text>
      <TextInput
        allowFontScaling={true}
        blurOnSubmit={true}
        maxLength={255}
        multiline={true}
        onChangeText={[Function]}
        onContentSizeChange={[Function]}
        onSubmitEditing={[Function]}
        placeholder="ie: Blue Lake, North Cascades, WA"
        rejectResponderTermination={true}
        returnKeyType="done"
        style={
          Array [
            Object {
              "backgroundColor": "#efefef",
              "borderColor": "gray",
              "borderRadius": 4,
              "borderWidth": 1,
              "minHeight": "8%",
              "paddingHorizontal": 15,
              "paddingVertical": 15,
            },
            Object {
              "paddingTop": 15,
            },
          ]
        }
        underlineColorAndroid="transparent"
        value=""
      />
      <Text
        style={
          Object {
            "fontSize": 15,
            "marginTop": 3,
          }
        }
      >
        Additional Notes (limit 255 characters)
      </Text>
      <TextInput
        allowFontScaling={true}
        blurOnSubmit={true}
        maxLength={255}
        multiline={true}
        onChangeText={[Function]}
        onContentSizeChange={[Function]}
        onSubmitEditing={[Function]}
        placeholder="ie. algae growing on glacial ice"
        rejectResponderTermination={true}
        returnKeyType="done"
        style={
          Array [
            Object {
              "backgroundColor": "#efefef",
              "borderColor": "gray",
              "borderRadius": 4,
              "borderWidth": 1,
              "minHeight": "8%",
              "paddingHorizontal": 15,
              "paddingVertical": 15,
            },
            Object {
              "paddingTop": 15,
            },
          ]
        }
        underlineColorAndroid="transparent"
        value=""
      />
      <Text
        style={
          Object {
            "fontSize": 15,
            "marginTop": 3,
          }
        }
      >
        Select Photos (limit 4)
      </Text>
      <View
        accessible={true}
        collapsable={false}
        focusable={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
      >
        <Text />
      </View>
    </View>
  </RCTScrollView>
</View>
`;
